### PR TITLE
AWS KMS encryption and decryption

### PR DESCRIPTION
**Pull request by gdxhsw**
# KMS class
- encrypt function
- decrypt function

# No Test

### CODE REVIEW

File: lib/kms.js [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+                        return reject(err)
+                    }
+                    return resolve(data.CiphertextBlob)
+                }
+            )
+        })
+    }
+
+    decrypt(ciphertextBlob) {
+        const Promise = require('bluebird')
```
Comment (wancw) [inline lib/kms.js]: 是因為考慮到 Lambda 的情況，所以才不把 require 放在檔案前面嗎？

File: package.json [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+  "keywords": [
+    "kms",
+    "aws",
+    "node"
+  ],
+  "dependencies": {
+    "bluebird": "3.3.5",
+    "aws-sdk": "2.7.20"
+  },
+  "author": "KKTV",
```
Comment (wancw) [inline package.json]: 慣例上好像會直接放 contributor （你）的資訊？

File: README.md [hunk 3]
```diff
@@ -1 +1,44 @@
...
+```js
+
+const KMS = new require('kms')
+const kms = new KMS(awsConfig)
+kms.decrypt(ciphertextBlob)
+    .then(plaintext => {
+        // plaintext is a Buffer object
+        console.log(plaintext.toString())
+    })
+
```
Comment (wancw) [inline README.md]: 要不要不論 input/output 加密過的是 Blob、明文是 String，這樣比較好記？

File: lib/kms.js (hunk 1 above)
Comment (gdxhsw) [inline lib/kms.js]: Lambda 是一個考量點，再來是也不一定每一個 function 都會用到 promise，所以就想說有用到再 require 就好

File: package.json (hunk 2 above)
Comment (gdxhsw) [inline package.json]: OK, 我改成 `GDX Wu <gdxwu@kktv.me>`

File: README.md (hunk 3 above)
Comment (gdxhsw) [inline README.md]: 因為這裡還考量到有可能加密的是 binary 資訊，不一定是明文

Comment (wancw) [inline README.md]: 那就都 Blob？一致的話，使用起來比較不花腦。

Comment (gdxhsw) [inline README.md]: Input 可以是 blob (Buffer) 喔，還是你是在說變數名稱會混淆用的人呢？

**Comment by badboy99tw** [comment]
type, commit log *dncrypt* function
要不要首字大寫或是改成 Add decrypt funciton

File: README.md (hunk 3 above)
Comment (gdxhsw) [inline README.md]: 我來改一下變數名稱好了，昨天在用 KMS API，也是對 input/output 不一致很困擾

**Comment by gdxhsw** [comment]
@badboy99tw type, commit log dncrypt function
typo 喔，type -> typo

**Comment by gdxhsw** [comment]
因為考量到 decrypt 其實是 function name 而不是一個句子的開頭...

Function \`decrypt\`

如何？

File: README.md (hunk 3 above)
Comment (wancw) [inline README.md]: 因為只看 README 會有點不太確定，多加個範例或是說明？
改變數名稱也行。

**Comment by wancw** [comment]
前面加個動詞， `Add` 或是 `Create` 之類的？

**Comment by badboy99tw** [comment]
我也是習慣動詞開頭（剛剛各種 typo...）

**Comment by gdxhsw** [comment]
OK 沒問題

File: README.md (hunk 3 above)
Comment (gdxhsw) [inline README.md]: 好，我也來看看怎麼修正 README 會比較好

File: lib/kms.js (hunk 1 above)
Comment (wancw) [inline lib/kms.js]: 了解！只是在想我們要不要慢慢歸納出慣例。

**Comment by gdxhsw** [comment]
[Open source aws-node-kms](https://trello.com/c/DKIYVDFH/2053-open-source-aws-node-kms)

**Comment by gdxhsw** [comment]
@wancw @badboy99tw 改好了！

* 修正 author 為 GDX Wu <gdxwu@kktv.me>
* 調整 commit message
* 調整 README 內容，說明 input/output 都是 Buffer object
* encrypt input/decrypt output 的變數名稱統一為 `bufferData`

File: README.md [hunk 4]
```diff
@@ -1 +1,49 @@
-# node-aws-kms
\ No newline at end of file
+node-aws-kms is a Node package that can helps you to encrypt/decrypt data in a easy way.
+
+# Encrypt data
+
+```js
+
+const KMS = new require('kms')
+const kms = new KMS(awsConfig)
```
Comment (wancw) [inline README.md]: Line 8 有 `new`，Line 7 的 `require` 前應該不用？ 下一段範例也是一樣的狀況。

**Comment by gdxhsw** [comment]
@wancw @badboy99tw 

I'm thinking that should I update AWS config by every `decrypt` and `encrypt` call.
The reason is, AWS config setting if global and different place may use different settings.

But if we use different settings in different place, the async call may causes problem...

File: README.md (hunk 4 above)
Comment (gdxhsw) [inline README.md]: 靠悲，打錯XDDDD

**Comment by wancw** [comment]
@gdxhsw 
In the example below, will `kms1` be affected by the `awsConfig2`? If so, I think we should make this library acts like AWS SDK, use 1 global config, instead of passing it as a constructor parameter.
 
```js
Aws.config.update(awsConfig1)
kms1 = new Aws.KMS()

Aws.config.update(awsConfig2)
kms2 = new Aws.KMS()
```

**Comment by badboy99tw** [comment]
góa kám-kah wancw kóng ê chin ū tō-lí.

**Comment by gdxhsw** [comment]
@wancw 

> In the example below, will kms1 be affected by the awsConfig2

The answer is **YES**

So in your oppinion, is it better to **NOT** setup AWS in node-aws-kms constructor?
I think it's convenience if we have it so the caller don't have to require aws-sdk and update the config before using it.

Both way are acceptable.

```js
const Aws = require('aws-sdk')
Aws.config.update(awsConfig)

const Kms = require('kms')
const kms = new Kms()
```

and

```js
const Kms = require('kms')
kms = new Kms(awsConfig)
```

**Comment by gdxhsw** [comment]
But, yes, the caller may think the awsConfig in kms won't effect global setting.

**Comment by wancw** [comment]
If the user is not familiar with AWS, he/she may not be aware a `new KMS(...)` will affect the objects created previously. Myabe we can offer our method to update the (AWS) config like the AWS SDK does.

```js
const KMS = require('kms')
KMS.updateConfig({...})

kms = new KMS()
```

On the other hand, if KMS module uses a globl config, is it really necessary to make it a class? How about just export oridinary functions?

**Comment by gdxhsw** [comment]
Agreed with you, OK I will modify it to just export ordinary functions.

**Comment by gdxhsw** [comment]
Modified!

* Just export ordinary functions.
* Change aws-sdk to peer dependencies.

**Comment by wancw** [comment]
Looks good to me!

**Pull request merged**